File: src/omega/GhostBrick.ts
import { hashOf } from "../utils/hash";
import { systemClock, type Clock } from "../utils/time";
import type { ShadowFrame } from "./contracts";

/**
 * BRICK_B_GHOST — Shadow Mirror Protocol.
 *
 * Maintains an atomic clone of the live state every cycle. The latest
 * shadow frame is the pointer-flip target the armor daemon swaps to
 * during sub-ms resurrection.
 */
export class GhostBrick {
  public static readonly IDENT = "BRICK_B_GHOST" as const;
  public static readonly STATE = "DORMANT_LIVE" as const;
  public static readonly TARGET_LATENCY_MS = 0.0001;

  private readonly frames: ShadowFrame[] = [];
  private cycleCounter = 0;

  constructor(
    private readonly clock: Clock = systemClock,
    private readonly nowMs: () => number = defaultNowMs,
    private readonly maxFrames: number = 8
  ) {}

  /** Atomically clone the supplied live state into a new shadow frame. */
  mirror(liveState: Readonly<Record<string, unknown>>): ShadowFrame {
    const start = this.nowMs();
    const cloned = deepClone(liveState);
    const frozen = deepFreeze(cloned);
    const latencyMs = Math.max(0, this.nowMs() - start);

    this.cycleCounter += 1;
    const frame: ShadowFrame = Object.freeze({
      cycle: this.cycleCounter,
      mirrorHash: hashOf(frozen),
      state: frozen,
      capturedAt: this.clock(),
      latencyMs
    });

    this.frames.push(frame);
    if (this.frames.length > this.maxFrames) {
      this.frames.splice(0, this.frames.length - this.maxFrames);
    }
    return frame;
  }

  /** Latest shadow frame, or undefined if no mirror has run yet. */
  latest(): ShadowFrame | undefined {
    return this.frames[this.frames.length - 1];
  }

  /** Return the bounded ring of shadow frames. */
  history(): readonly ShadowFrame[] {
    return this.frames.slice();
  }

  /** Pointer-flip: hand back the most recent frame for resurrection. */
  pointerFlip(): ShadowFrame {
    const frame = this.latest();
    if (!frame) {
      throw new Error("GhostBrick: pointer flip requested before any mirror cycle.");
    }
    return frame;
  }
}

function defaultNowMs(): number {
  if (typeof globalThis !== "undefined") {
    const perf = (globalThis as { performance?: { now(): number } }).performance;
    if (perf && typeof perf.now === "function") return perf.now();
  }
  return Date.now();
}

function deepClone<T>(value: T): T {
  if (value === null || typeof value !== "object") return value;
  if (Array.isArray(value)) {
    return value.map((item) => deepClone(item)) as unknown as T;
  }
  const out: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
    out[k] = deepClone(v);
  }
  return out as T;
}

function deepFreeze<T>(value: T): T {
  if (value === null || typeof value !== "object") return value;
  for (const key of Object.keys(value as object)) {
    const child = (value as Record<string, unknown>)[key];
    if (child && typeof child === "object") deepFreeze(child);
  }
  return Object.freeze(value);
}

File: sovereign_stitch/ghost_brick.py
"""
BRICK_B_GHOST — Shadow Mirror Protocol.

Maintains an atomic clone of the live state every cycle.  The latest
shadow frame is the pointer-flip target the armor daemon swaps to
during sub-ms resurrection.
"""

import hashlib
import json
import time
from collections import deque
from copy import deepcopy
from typing import Any, Deque, Dict, Optional


def _stable_json(data: Any) -> str:
    return json.dumps(data, sort_keys=True, separators=(",", ":"))


def _hash_of(data: Any) -> str:
    if isinstance(data, (dict, list)):
        data = _stable_json(data)
    if isinstance(data, str):
        data = data.encode()
    return hashlib.sha256(data).hexdigest()


class ShadowFrame:
    """An immutable snapshot of a single mirror cycle."""

    def __init__(self, cycle: int, state: Dict[str, Any], latency_ms: float) -> None:
        self.cycle = cycle
        self.state = deepcopy(state)
        self.mirror_hash = _hash_of(_stable_json(state))
        self.captured_at = time.time()
        self.latency_ms = latency_ms

    def __repr__(self) -> str:
        return f"ShadowFrame(cycle={self.cycle}, hash={self.mirror_hash[:12]}…, latency={self.latency_ms:.4f}ms)"


class GhostBrick:
    """BRICK_B_GHOST — Shadow Mirror Protocol."""

    IDENT = "BRICK_B_GHOST"
    STATE = "DORMANT_LIVE"
    TARGET_LATENCY_MS = 0.0001

    def __init__(self, max_frames: int = 8) -> None:
        self._frames: Deque[ShadowFrame] = deque(maxlen=max_frames)
        self._cycle = 0

    def mirror(self, live_state: Dict[str, Any]) -> ShadowFrame:
        """Atomically clone the supplied live state into a new shadow frame."""
        start = time.perf_counter()
        cloned = deepcopy(live_state)
        latency_ms = (time.perf_counter() - start) * 1000.0

        self._cycle += 1
        frame = ShadowFrame(self._cycle, cloned, latency_ms)
        self._frames.append(frame)
        return frame

    def latest(self) -> Optional[ShadowFrame]:
        """Latest shadow frame, or None if no mirror has run yet."""
        if not self._frames:
            return None
        return self._frames[-1]

    def history(self) -> list:
        """Return the bounded ring of shadow frames."""
        return list(self._frames)

    def pointer_flip(self) -> ShadowFrame:
        """Return the most recent frame for resurrection."""
        frame = self.latest()
        if frame is None:
            raise RuntimeError("GhostBrick: pointer flip requested before any mirror cycle.")
        return frame
